accelerated_scan/warp.cuh

```diff
@@ -342,4 +344,371 @@ warpscan_forward(const at::Tensor &gates, const at::Tensor &tokens, const at::Te
         TORCH_CHECK(false && "Unsupported tensor dtype: expecting bfloat16, float16 or float32");
     }
     return out;
-}
\ No newline at end of file
+}
+
+template <typename Tuple, int offset>
+__device__ Tuple load_shifted_tuple(const Tuple* ptr, int index, int minIdx, int maxIdx) {
+    using weight_t = typename Tuple::Type;
+
+    const weight_t* rawPtr = reinterpret_cast<const weight_t *>(ptr);
+    Tuple x;
+    for (int i = 0; i < Tuple::Size; i++) {
+        const int idx = index * Tuple::Size + i + offset;
+        if (idx >= minIdx * Tuple::Size && idx < maxIdx * Tuple::Size) {
+            x.data[i] = rawPtr[idx];
+        } else {
+            x.data[i] = 0.0;
```

Comment:
When offset is 1 we need to pad with values of 1 — I'm a bit confused why it doesn't show up on your test. Would like to play with this.

```python
padded_shifted_gates = torch.cat([gates, torch.ones_like(gates[:, :, :1])], dim=-1)[:, :, 1:].contiguous()
d_states = scan_forward(padded_shifted_gates, grad_output, reverse=True)
padded_outputs = torch.cat([torch.zeros_like(states[:, :, :1]), states], dim=-1)[:, :, :-1]
```

Reply:
I think it's because the first gate in a scan is unused (since the -1'th acc is 0). So it won't matter what we pad.